### ISSUE TITLE
Clone Slider

### ISSUE THREAD

**Issue by brewmasterhal** [ousnius/BodySlide-and-Outfit-Studio#516]
I recently attempted to make a copy of an existing slider (AreolaSize) so that I could split it into left and right, and control them independently.

With the existing UI, it's easy to mask the verts, clear the mask from one side of the body, and then clear the unmasked slider data to make a slider only apply to one side.  There is not an easy way to make a copy of a slider so that I can perform this procedure twice, and have both left-side and right-side sliders.

I was unable to find any method within the existing application to accomplish this, and no resources online suggesting a method either.

I initially attempted to use the Export Slider Data and Import Slider Data menu options, but this simply over-wrote the AreolaSize slider data instead of copying it to the new slider I had created and placed in edit mode.

Next I attempted to modify the .osp file to create two new sliders and direct them to read the same data from the 3BAAreolaSize region of the .osd file.  This allowed the new "left" and "right" sliders to work, but they were merely aliases of the original slider.  Making modifications to any of the three modified all of them.

Ultimately to get it to work, I downloaded the source code and looked for the .osd format parser.  Between the contents of DiffData.cpp and the relevant .osd file open in a hex editor, I was able to reverse-engineer the format enough to copy the 3BAAreolaSize data twice, rename the new regions, modify the relevant bytes to correct for String length and number of data blocks, modify the .osp file to create new sliders that point to the copied/renamed data, and create clones of the desired slider.

For anyone else who is interested, the OSD file format is as follows:
- 4-byte header containing 0x00, D, S, O
- 4-byte version number in little-endian (0x01000000)
- 4-byte number of data blocks in little-endian (0xeb000000 is the default for 3BA I believe)
- Beginning of the first block - Each block consists of:
-   1-byte length of block name string
-   X bytes of string
-   2-byte number of "dataDiff" structs
-   Y * 14 (hex 0xE) bytes of data

The DiffStruct type consists of a uint16 and a Vector3, presumably being an index of a vertex in the model file, and an offset indicating where the vertex moves to (relative) when the slider is at 100%.

If there is a sane way to accomplish cloning a slider, I would very much appreciate being informed of it.  If not, then I humbly request that one be added.

I realize that I may be the first user to ever attempt such a thing (and therefore this is likely a very low priority), but it seems to me that such a feature would be quite useful.

**Comment by ousnius** [comment]
1. Slider -> Export -> As .BSD file
2. Slider -> New Slider
3. Slider -> Properties -> Turn it into a UV slider
4. Slider -> Import -> From .BSD file

When you have the source code, it's not reverse-engineering. ;)

**Comment by brewmasterhal** [comment]
Tested your procedure with a non-UV slider, works just fine.  Is there something special about the .BSD file format?  Seems to be the only one that exports a single slider's data in isolation.

So clearly all the requisite functionality exists.  I'll try to dive into the source again and see what I can do about kluging together a script that does the mechanical bits.  Hardest part should be _not_ creating an unnecessary bsd file.  If I get it to work, I'll send you a pull request.

...

I mean yeah, it's not, like, hard to do what I did.  I could always go full hacker mode and decompile the bin with IDA, or set some debug hooks and check the x64 - but why, when I have your source?  Seems like a semantic difference to me.
